### PR TITLE
expression: implement vectorized evaluation for `builtinArithmeticMinusIntSig`

### PR DESCRIPTION
**Pull request by js00070**
PCP #12103 

### What problem does this PR solve? 
implement vectorized `builtinArithmeticMinusIntSig` mentioned in https://github.com/pingcap/tidb/issues/12103

### What is changed and how it works?
get 10 times faster
```
BenchmarkVectorizedBuiltinArithmeticFunc/builtinArithmeticMinusIntSig-VecBuiltinFunc-4                            179317              6536 ns/op               0 B/op           0 allocs/op
BenchmarkVectorizedBuiltinArithmeticFunc/builtinArithmeticMinusIntSig-NonVecBuiltinFunc-4                          17035             68626 ns/op               0 B/op           0 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_arithmetic_vec.go [hunk 1]
```diff
@@ -182,11 +182,183 @@ func (b *builtinArithmeticMinusDecimalSig) vecEvalDecimal(input *chunk.Chunk, re
...
+		if lh < 0 || (lh > math.MaxInt64) {
+			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
+		}
+
+		if rh < 0 || (rh > math.MaxInt64) {
+			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
+		}
+
+		if (lh > 0 && -rh > math.MaxInt64-lh) || (lh < 0 && -rh < math.MinInt64-lh) {
+			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s + %s)", b.args[0].String(), b.args[1].String()))
```
Comment (wshwsh12) [inline expression/builtin_arithmetic_vec.go]: s/  BIGINT UNSIGNED /BIGINT

File: expression/builtin_arithmetic_vec.go [hunk 2]
```diff
@@ -182,11 +182,183 @@ func (b *builtinArithmeticMinusDecimalSig) vecEvalDecimal(input *chunk.Chunk, re
...
+			continue
+		}
+		lh, rh := lhi64s[i], rhi64s[i]
+
+		if lh < 0 || (lh > math.MaxInt64) {
+			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
+		}
+
+		if (lh > 0 && -rh > math.MaxInt64-lh) || (lh < 0 && -rh < math.MinInt64-lh) {
+			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s + %s)", b.args[0].String(), b.args[1].String()))
```
Comment (wshwsh12) [inline expression/builtin_arithmetic_vec.go]: ditto.

File: expression/builtin_arithmetic_vec.go [hunk 3]
```diff
@@ -182,11 +182,183 @@ func (b *builtinArithmeticMinusDecimalSig) vecEvalDecimal(input *chunk.Chunk, re
...
+			continue
+		}
+		lh, rh := lhi64s[i], rhi64s[i]
+
+		if rh < 0 || (rh > math.MaxInt64) {
+			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
+		}
+
+		if (lh > 0 && -rh > math.MaxInt64-lh) || (lh < 0 && -rh < math.MinInt64-lh) {
+			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s + %s)", b.args[0].String(), b.args[1].String()))
```
Comment (wshwsh12) [inline expression/builtin_arithmetic_vec.go]: ditto.

File: expression/builtin_arithmetic_vec.go [hunk 4]
```diff
@@ -182,11 +182,183 @@ func (b *builtinArithmeticMinusDecimalSig) vecEvalDecimal(input *chunk.Chunk, re
...
+}
+func (b *builtinArithmeticMinusIntSig) minusSS(result *chunk.Column, lhi64s, rhi64s, resulti64s []int64) error {
+	for i := 0; i < len(lhi64s); i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		lh, rh := lhi64s[i], rhi64s[i]
+
+		if (lh > 0 && -rh > math.MaxInt64-lh) || (lh < 0 && -rh < math.MinInt64-lh) {
+			return types.ErrOverflow.GenWithStackByArgs("BIGINT", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
```
Comment (wshwsh12) [inline expression/builtin_arithmetic_vec.go]: ditto.

Comment (js00070) [reply expression/builtin_arithmetic_vec.go]: yes, this struct is perfect, I will change the code.

**Comment by js00070** [comment]
> Why not use the same process control struct in non-vec function. Now it is too difficult to review and check the correctness.
> 
> ```
> forceToSigned := s.ctx.GetSessionVars().SQLMode.HasNoUnsignedSubtractionMode()
> isLHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag)
> isRHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
> 
> if forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag) {
> 	for i := 0;i < n;i++ {
> 		if lhi64s[i] < 0 || (lhi64s[i] > math.MaxInt64) {
> 			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
> 		}
> 	}
> }
> if forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag) {
> ...
> }
> switch {
> ...
> }
> for i := 0;i < n;i++ {
> resulti64s[i] = lhi64s[i]-rhi64i[i]
> }
> return nil
> ```
> 
> I think this struct is good. Consider it?

I take a try and find a potential problem in this code structrue.

in the non-vec function, if `args[0]` or `args[1]` is `Null`, the function will return `IsNull` immediately and will not check the following `BIGINT` condition.

but in vectorized version, if we use this code structure, we have to add `if buf.IsNull(i)` in every `for` loop of all check step to get the exactly the same process of non-vec version. If we don't add, the behavior may be different in some cases? but adding so many `IsNull` may cause performance degradation in large amount of rows.

**Comment by wshwsh12** [comment]
> > Why not use the same process control struct in non-vec function. Now it is too difficult to review and check the correctness.
> > ```
> > forceToSigned := s.ctx.GetSessionVars().SQLMode.HasNoUnsignedSubtractionMode()
> > isLHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag)
> > isRHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
> > 
> > if forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag) {
> > 	for i := 0;i < n;i++ {
> > 		if lhi64s[i] < 0 || (lhi64s[i] > math.MaxInt64) {
> > 			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
> > 		}
> > 	}
> > }
> > if forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag) {
> > ...
> > }
> > switch {
> > ...
> > }
> > for i := 0;i < n;i++ {
> > resulti64s[i] = lhi64s[i]-rhi64i[i]
> > }
> > return nil
> > ```
> > 
> > 
> > I think this struct is good. Consider it?
> 
> I take a try and find a potential problem in this code structrue.
> 
> in the non-vec function, if `args[0]` or `args[1]` is `Null`, the function will return `IsNull` immediately and will not check the following `BIGINT` condition.
> 
> but in vectorized version, if we use this code structure, we have to add `if buf.IsNull(i)` in every `for` loop of all check step to get the exactly the same process of non-vec version. If we don't add, the behavior may be different in some cases? but adding so many `IsNull` may cause performance degradation in large amount of rows.

How about introduce a int64 slice to record the index which row is null?
```
MergeNull
indexes := make([]int64,n)
for i := 0;i < n;i++{
    if !isnull(i) indexes = append(indexes,i)
}

if check {
    for _, idx := range indexes {
        if check {
            return err   
        }
    }
}
```
I don't know how much the performance degradation in so many `IsNull`, we can have a test.
If the result is not really good, maybe we need abandon the old struct and try a new struct(like you code now).

**Comment by js00070** [comment]
> > > Why not use the same process control struct in non-vec function. Now it is too difficult to review and check the correctness.
> > > ```
> > > forceToSigned := s.ctx.GetSessionVars().SQLMode.HasNoUnsignedSubtractionMode()
> > > isLHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag)
> > > isRHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
> > > 
> > > if forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag) {
> > > 	for i := 0;i < n;i++ {
> > > 		if lhi64s[i] < 0 || (lhi64s[i] > math.MaxInt64) {
> > > 			return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
> > > 		}
> > > 	}
> > > }
> > > if forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag) {
> > > ...
> > > }
> > > switch {
> > > ...
> > > }
> > > for i := 0;i < n;i++ {
> > > resulti64s[i] = lhi64s[i]-rhi64i[i]
> > > }
> > > return nil
> > > ```
> > > 
> > > 
> > > I think this struct is good. Consider it?
> > 
> > 
> > I take a try and find a potential problem in this code structrue.
> > in the non-vec function, if `args[0]` or `args[1]` is `Null`, the function will return `IsNull` immediately and will not check the following `BIGINT` condition.
> > but in vectorized version, if we use this code structure, we have to add `if buf.IsNull(i)` in every `for` loop of all check step to get the exactly the same process of non-vec version. If we don't add, the behavior may be different in some cases? but adding so many `IsNull` may cause performance degradation in large amount of rows.
> 
> How about introduce a int64 slice to record the index which row is null?
> 
> ```
> MergeNull
> indexes := make([]int64,n)
> for i := 0;i < n;i++{
>     if !isnull(i) indexes = append(indexes,i)
> }
> 
> if check {
>     for _, idx := range indexes {
>         if check {
>             return err   
>         }
>     }
> }
> ```
> 
> I don't know how much the performance degradation in so many `IsNull`, we can have a test.
> If the result is not really good, maybe we need abandon the old struct and try a new struct(like you code now).

I benchtested all the code versions we mentioned and got the compare results.

I runed the benchtest using the following commands
```
go test -v -benchmem -bench=BenchmarkVectorizedBuiltinArithmeticFunc -count=20 -run=BenchmarkVectorizedBuiltinArithmeticFunc | tee mycode.txt

go test -v -benchmem -bench=BenchmarkVectorizedBuiltinArithmeticFunc -count=20 -run=BenchmarkVectorizedBuiltinArithmeticFunc | tee new1.txt

go test -v -benchmem -bench=BenchmarkVectorizedBuiltinArithmeticFunc -count=20 -run=BenchmarkVectorizedBuiltinArithmeticFunc | tee new2.txt
```

and the result is
```bash
benchstat mycode.txt new1.txt
name                                                                               old time/op    new time/op    delta
VectorizedBuiltinArithmeticFunc/builtinArithmeticMinusIntSig-VecBuiltinFunc-12       4.88µs ± 1%    6.91µs ± 1%  +41.76%  (p=0.000 n=20+20)

benchstat mycode.txt new2.txt
name                                                                               old time/op    new time/op     delta
VectorizedBuiltinArithmeticFunc/builtinArithmeticMinusIntSig-VecBuiltinFunc-12       4.88µs ± 1%    36.92µs ± 2%  +657.11%  (p=0.000 n=20+20)

```

the `mycode.txt` is the benchtest result of my original code

the `new1.txt` is the benchtest result of the code which uses tons of `IsNull(i)`, code as follow:
```go
	forceToSigned := b.ctx.GetSessionVars().SQLMode.HasNoUnsignedSubtractionMode()
	isLHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(b.args[0].GetType().Flag)
	isRHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(b.args[1].GetType().Flag)

	if forceToSigned && mysql.HasUnsignedFlag(b.args[0].GetType().Flag) {
		for i := 0; i < n; i++ {
			if result.IsNull(i) {
				continue
			}
			if lhi64s[i] < 0 || (lhi64s[i] > math.MaxInt64) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	}

	if forceToSigned && mysql.HasUnsignedFlag(b.args[1].GetType().Flag) {
		for i := 0; i < n; i++ {
			if result.IsNull(i) {
				continue
			}
			if rhi64s[i] < 0 || (rhi64s[i] > math.MaxInt64) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	}

	switch {
	case isLHSUnsigned && isRHSUnsigned:
		for i := 0; i < n; i++ {
			if result.IsNull(i) {
				continue
			}
			if uint64(lhi64s[i]) < uint64(rhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	case isLHSUnsigned && !isRHSUnsigned:
		for i := 0; i < n; i++ {
			if result.IsNull(i) {
				continue
			}
			if rhi64s[i] >= 0 && uint64(lhi64s[i]) < uint64(rhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
			if rhi64s[i] < 0 && uint64(lhi64s[i]) > math.MaxUint64-uint64(-rhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	case !isLHSUnsigned && isRHSUnsigned:
		for i := 0; i < n; i++ {
			if result.IsNull(i) {
				continue
			}
			if uint64(lhi64s[i]-math.MinInt64) < uint64(rhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	case !isLHSUnsigned && !isRHSUnsigned:
		for i := 0; i < n; i++ {
			if result.IsNull(i) {
				continue
			}
			if (lhi64s[i] > 0 && -rhi64s[i] > math.MaxInt64-lhi64s[i]) || (lhi64s[i] < 0 && -rhi64s[i] < math.MinInt64-lhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	}
	for i := 0; i < n; i++ {
		if result.IsNull(i) {
			continue
		}
		resulti64s[i] = lhi64s[i] - rhi64s[i]
	}

	return err
```

the `new2.txt` is the benchtest result of the code which uses a `indexes` slice to replace `IsNull` just like you mentioned, code as follow:

```go
	indexes := make([]int, n)
	for i := 0; i < n; i++ {
		if !result.IsNull(i) {
			indexes = append(indexes, int(i))
		}
	}

	if forceToSigned && mysql.HasUnsignedFlag(b.args[0].GetType().Flag) {
		for _, i := range indexes {
			if lhi64s[i] < 0 || (lhi64s[i] > math.MaxInt64) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	}

	if forceToSigned && mysql.HasUnsignedFlag(b.args[1].GetType().Flag) {
		for _, i := range indexes {
			if rhi64s[i] < 0 || (rhi64s[i] > math.MaxInt64) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	}

	switch {
	case isLHSUnsigned && isRHSUnsigned:
		for _, i := range indexes {
			if uint64(lhi64s[i]) < uint64(rhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	case isLHSUnsigned && !isRHSUnsigned:
		for _, i := range indexes {
			if rhi64s[i] >= 0 && uint64(lhi64s[i]) < uint64(rhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
			if rhi64s[i] < 0 && uint64(lhi64s[i]) > math.MaxUint64-uint64(-rhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	case !isLHSUnsigned && isRHSUnsigned:
		for _, i := range indexes {
			if uint64(lhi64s[i]-math.MinInt64) < uint64(rhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	case !isLHSUnsigned && !isRHSUnsigned:
		for _, i := range indexes {
			if (lhi64s[i] > 0 && -rhi64s[i] > math.MaxInt64-lhi64s[i]) || (lhi64s[i] < 0 && -rhi64s[i] < math.MinInt64-lhi64s[i]) {
				return types.ErrOverflow.GenWithStackByArgs("BIGINT", fmt.Sprintf("(%s - %s)", b.args[0].String(), b.args[1].String()))
			}
		}
	}
	for _, i := range indexes {
		resulti64s[i] = lhi64s[i] - rhi64s[i]
	}
```

We can see my origin code is the fastest. I think we should keep it.

BTW, it's pretty werid that the `new2` version is so much slower than the `new1` version. Can anyone tell the reason? I guess it may be about the CPU hardware optimization but not really sure.

**Comment by wshwsh12** [comment]
@js00070  OK , I also think we should keep it. Please address comments that `s/ BIGINT UNSIGNED /BIGINT`. And i will review your origin code again.

**Comment by js00070** [comment]
> @js00070 OK , I also think we should keep it. Please address comments that `s/ BIGINT UNSIGNED /BIGINT`. And i will review your origin code again.

It seems that we should use `BIGINT UNSIGNED` because in the non-vec function, it alse used `BIGINT UNSIGNED`. Maybe we should keep the vec version's behaviour same as the non-vec version?

**Comment by wshwsh12** [comment]
> 
> It seems that we should use `BIGINT UNSIGNED` because in the non-vec function, it alse used `BIGINT UNSIGNED`. Maybe we should keep the vec version's behaviour same as the non-vec version?

The check `(lh > 0 && -rh > math.MaxInt64-lh) || (lh < 0 && -rh < math.MinInt64-lh)` in your code is different from `(a > 0 && -b > math.MaxInt64-a) || (a < 0 && -b < math.MinInt64-a)` in non-vec function?  In non-vec function, it uses `BIGINT`.

**Comment by js00070** [comment]
> > It seems that we should use `BIGINT UNSIGNED` because in the non-vec function, it alse used `BIGINT UNSIGNED`. Maybe we should keep the vec version's behaviour same as the non-vec version?
> 
> The check `(lh > 0 && -rh > math.MaxInt64-lh) || (lh < 0 && -rh < math.MinInt64-lh)` in your code is different from `(a > 0 && -b > math.MaxInt64-a) || (a < 0 && -b < math.MinInt64-a)` in non-vec function? In non-vec function, it uses `BIGINT`.

yes, you are right. I have fixed it now.

**Comment by wshwsh12** [comment]
The logic is same with non-vec function, but the code struct is really bad.  @qw4990  PATL . And any idea?

**Pull request merged**